src/client/SOAP_client.ts

```diff
@@ -1,22 +1,121 @@
-        this.port = "8085";
-        this.url = "http://build-oracc.museum.upenn.edu"
-        this.method = "POST";
-        this.atf_path = atf_path;
-        this.atf_filename = path.basename(atf_path);
+        this.atf_filename = atf_name;
+        this.atf_project = atf_project;
+        this.atf_text = atf_text;
+    }
+
+    async executeCommand(command: ServerAction): Promise<ServerResult> {
+        try {
+            this.responseID = await this.sendInitialRequest(command);
+            log('debug', `Got response code ${this.responseID}`);
+            await this.serverComplete();
+            log('info', `Request ${this.responseID} is done.`);
+            const finalResult = await this.retrieveReponse();
+            return new ServerResult(finalResult.get('oracc.log'),
+                                    finalResult.get('request.log'));
+        } catch (err) {
+            log('error', `Error during communication with server: ${err}`);
+            vscode.window.showErrorMessage(
+                "Problem getting response from server, see log for details.");
+            return;  // Should return an empty result?
+        }
+    }
+
+    async sendInitialRequest(command): Promise<string> {
+        const fullMessage = new MultipartMessage(command, this.atf_filename,
+                                                 this.atf_project, this.atf_text);
+        log('debug', fullMessage.toString());
+        const body = fullMessage.getBody();
+
+        const headers = fullMessage.getHeaders();
 
+        const options = {
+            method: 'POST',
+            headers: headers,
+            body: body
+        }
+
+        try {
+            const response = await fetch(`${this.url}:${this.port}`, options);
+            if (response.status != 200) {
+                throw `Request failed! Status: ${response.status}`;
+            }
+            return getResponseCode(await response.text());
+        } catch (err) {
+            log('error', `Problem getting response from server: ${err}`);
+            throw err;
+        }
+    }
+
+    async serverComplete(): Promise<void> {
+        let attempts = 0;
+        const max_attempts = 10;
+        while (attempts < max_attempts) {
+            attempts += 1;
+            const response = await fetch(`${this.url}/p/${this.responseID}`);
+            const text = await response.text();
+            switch (text.trim()) { // Message includes a trailing new line character
+                case 'done': // done processing
+                    return;
+                case 'err_stat':
+                    log('error', `Received err_stat for request ${this.responseID}.`);
+                    throw `The server encountered an error while validating.
+                        Please contact the Oracc server admin to look into this problem.`;
+                case 'run':
+                    if (attempts < max_attempts) {
+                        log('info', 'Server working on request.');
```

Comment:
Does this mean that the user will see "Server working on request" as many times as attempts there are? I don't know how many of those it normally takes until the server finished processing the request, but I'd log the attempt number in the debug log maybe, and just the "Server working on request" for the users to see just once. Or even not at all.

Reply:
I was going by what Nammu does (it prints every time on the console), but you're right that it doesn't show it on the main window. This will only show up to ten times (we'll time out after that).

I wanted to have something that shows that it's not stuck, but your suggestion sounds better, thanks!